internal/datastore/memdb/memdb.go

```diff
@@ -348,93 +34,199 @@ func NewMemdbDatastore(
+	})
+
+	if revIndex == len(mdb.revisions) {
+		revIndex = len(mdb.revisions) - 1
 	}
 
-	newChangelogEntry := &transaction{
-		id:        newTransactionID,
-		timestamp: uint64(time.Now().UnixNano()),
+	snapshotRevision := mdb.revisions[revIndex].revision
+	roTxn := mdb.revisions[revIndex].db.Txn(false)
+
+	txSrc := func() (*memdb.Txn, error) {
+		return roTxn, nil
 	}
 
-	if err := txn.Insert(tableTransaction, newChangelogEntry); err != nil {
-		return 0, err
+	return &memdbReader{noopTryLocker{}, txSrc, snapshotRevision, nil}
+}
+
+func (mdb *memdbDatastore) ReadWriteTx(
+	ctx context.Context,
+	f datastore.TxUserFunc,
+) (datastore.Revision, error) {
+	for i := 0; i < 10; i++ {
+		var tx *memdb.Txn
+		createTxOnce := sync.Once{}
+		txSrc := func() (*memdb.Txn, error) {
+			var err error
+			createTxOnce.Do(func() {
+				mdb.Lock()
+				defer mdb.Unlock()
+
+				if mdb.activeWriteTxn != nil {
+					err = errSerialization
+					return
+				}
+
+				tx = mdb.db.Txn(true)
+				tx.TrackChanges()
+				mdb.activeWriteTxn = tx
+			})
+
+			return tx, err
+		}
+
+		newRevision := revisionFromTimestamp(time.Now().UTC())
+
+		rwt := &memdbReadWriteTx{memdbReader{&sync.Mutex{}, txSrc, datastore.NoRevision, nil}, newRevision}
+		if err := f(ctx, rwt); err != nil {
+			mdb.Lock()
+			if tx != nil {
+				tx.Abort()
+				mdb.activeWriteTxn = nil
+			}
+
+			// If the error was a serialization error, retry the transaction
+			if errors.Is(err, errSerialization) {
+				mdb.Unlock()
+
+				// If we don't sleep here, we run out of retries instantaneously
+				time.Sleep(1 * time.Millisecond)
+				continue
+			}
+			defer mdb.Unlock()
+
+			// We *must* return the inner error unmodified in case it's not an error type
+			// that supports unwrapping (e.g. gRPC errors)
+			return datastore.NoRevision, err
+		}
+
+		mdb.Lock()
+		defer mdb.Unlock()
+
+		// Record the changes that were made
+		newChanges := datastore.RevisionChanges{
+			Revision: newRevision,
+			Changes:  nil,
+		}
+		if tx != nil {
+			for _, change := range tx.Changes() {
+				if change.Table == "relationship" {
+					if change.After != nil {
+						newChanges.Changes = append(newChanges.Changes, &corev1.RelationTupleUpdate{
+							Operation: corev1.RelationTupleUpdate_TOUCH,
+							Tuple:     change.After.(*relationship).RelationTuple(),
+						})
+					}
+					if change.After == nil && change.Before != nil {
+						newChanges.Changes = append(newChanges.Changes, &corev1.RelationTupleUpdate{
+							Operation: corev1.RelationTupleUpdate_DELETE,
+							Tuple:     change.Before.(*relationship).RelationTuple(),
+						})
+					}
+				}
+			}
+
+			change := &changelog{
+				revisionNanos: newRevision.IntPart(),
+				changes:       newChanges,
+			}
+			if err := tx.Insert(tableChangelog, change); err != nil {
+				return datastore.NoRevision, fmt.Errorf("error writing changelog: %w", err)
+			}
+
+			tx.Commit()
+		}
+		mdb.activeWriteTxn = nil
```

Comment:
newChanges can be defined in the conditional below.

The whole thing could be made a little more clear with a refactor:

```go
if mdb.activeWriteTxn != nil {
	insertChanges(tx, newRevision)
	mdb.activeWriteTxn.Commit()
}
mdb.activeWriteTxn = nil
```